compliance/virtualmachines/relay/relay.go

```diff
@@ -81,38 +115,56 @@ func (r *Relay) Run() error {
 	}()
 
 	for {
+		if err := r.vsockServer.acquireSemaphore(r.ctx); err != nil {
+			if r.ctx.Err() != nil {
+				log.Info("Stopping virtual machine relay")
+				return r.ctx.Err()
+			}
+
+			log.Warnf("Failed to acquire semaphore to handle connection after %v: %v",
+				r.vsockServer.maxSemaphoreWaitTime, err,
+			)
+
+			continue
+		}
+
 		// Accept() is blocking, but it will return when ctx is cancelled and the above goroutine calls r.vsockServer.Stop()
```

Comment:
Is there an incoming connection queue defined for the listener ? If so, there is a chance that agents connection attempts pile in that queue and that some get `CONN_REFUSED` errors while other wait for a semaphore slot to become free.

Reply:
Yes, there is an accept queue at the OS level. I have added details to the PR description! The TL;DR is that the unaccepted connections pile up there up to `unix.SOMAXCONN` which is 128, after which new connections will be refused. The queued ones will be handled as new slots become available.